File: thrifty-schema/src/main/java/com/microsoft/thrifty/schema/parser/ThriftyParserPlugins.java
package com.microsoft.thrifty.schema.parser;

import java.util.UUID;

/**
 * Utility class to inject handlers to certain standard Thrifty operations.
 */
public final class ThriftyParserPlugins {

    private static final UUIDProvider DEFAULT_UUID_PROVIDER = UUID::randomUUID;
    private static volatile UUIDProvider uuidProvider = DEFAULT_UUID_PROVIDER;

    /**
     * Prevents changing the plugins.
     */
    private static volatile boolean lockdown;

    /**
     * Prevents changing the plugins from then on.
     * <p>
     * This allows container-like environments to prevent client messing with plugins.
     */
    public static void lockdown() {
        lockdown = true;
    }

    /**
     * Returns true if the plugins were locked down.
     *
     * @return true if the plugins were locked down
     */
    public static boolean isLockdown() {
        return lockdown;
    }

    /**
     * @param uuidProvider the provider to use for generating {@link UUID}s for elements.
     */
    public static void setUUIDProvider(UUIDProvider uuidProvider) {
        if (lockdown) {
            throw new IllegalStateException("Plugins can't be changed anymore");
        }
        ThriftyParserPlugins.uuidProvider = uuidProvider;
    }

    /**
     * @return a {@link UUID} as dictated by {@link #uuidProvider}. Default is random UUIDs.
     */
    public static UUID createUUID() {
        return uuidProvider.call();
    }

    public static void reset() {
        uuidProvider = DEFAULT_UUID_PROVIDER;
    }

    private ThriftyParserPlugins() {
        // No instances.
    }

    /**
     * A simple provider interface for creating {@link UUID}s.
     */
    public interface UUIDProvider {

        /**
         * @return a {@link UUID}.
         */
        UUID call();
    }

}

File: thrifty-schema/src/main/kotlin/com/microsoft/thrifty/schema/parser/ThriftyParserPlugins.kt
/*
 * Thrifty
 *
 * Copyright (c) Microsoft Corporation
 *
 * All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the License);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING
 * WITHOUT LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE,
 * FITNESS FOR A PARTICULAR PURPOSE, MERCHANTABLITY OR NON-INFRINGEMENT.
 *
 * See the Apache Version 2.0 License for specific language governing permissions and limitations under the License.
 */
package com.microsoft.thrifty.schema.parser

import java.util.UUID

object ThriftyParserPlugins {
    private val DEFAULT_UUID_PROVIDER: UUIDProvider = object : UUIDProvider {
        override fun call(): UUID = UUID.randomUUID()
    }

    @Volatile
    private var uuidProvider = DEFAULT_UUID_PROVIDER

    /**
     * Prevents changing the plugins.
     */
    @Volatile
    private var lockdown: Boolean = false

    /**
     * Prevents changing the plugins from then on.
     *
     *
     * This allows container-like environments to prevent client messing with plugins.
     */
    fun lockdown() {
        lockdown = true
    }

    /**
     * Returns true if the plugins were locked down.
     *
     * @return true if the plugins were locked down
     */
    fun isLockdown(): Boolean {
        return lockdown
    }

    /**
     * @param uuidProvider the provider to use for generating [UUID]s for elements.
     */
    fun setUUIDProvider(uuidProvider: UUIDProvider) {
        if (lockdown) {
            throw IllegalStateException("Plugins can't be changed anymore")
        }
        ThriftyParserPlugins.uuidProvider = uuidProvider
    }

    /**
     * @return a [UUID] as dictated by [uuidProvider]. Default is random UUIDs.
     */
    fun createUUID(): UUID {
        return uuidProvider.call()
    }

    fun reset() {
        uuidProvider = DEFAULT_UUID_PROVIDER
    }

    /**
     * A simple provider interface for creating [UUID]s.
     */
    interface UUIDProvider {

        /**
         * @return a [UUID].
         */
        fun call(): UUID
    }
}